Add library support to execute cephadm-ansible module playbooks

vamahaja opened the pull request:
Signed-off-by: Vaibhav Mahajan <vamahaja@redhat.com>

# Description
Add support to execute cephadm ansible wrapper playbooks.

vamahaja commented:
Testrun Results - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-CVEJFJ/

sunilangadi2 commented on playbooks/bootstrap-cluster.yml:
```diff
@@ -0,0 +1,15 @@
+- name: Test 'cephadm_bootstrap' module
+  hosts: admin
+  gather_facts: false
+  become: true
```
what is the actual usage of `become` can we have more user friendly arg names?

vamahaja replied:
`become` is ansible parameter to execute task with `root` previllage. Refer - https://docs.ansible.com/ansible/latest/user_guide/become.html#id2

vamahaja commented on playbooks/bootstrap-cluster.yml:
```diff
@@ -0,0 +1,15 @@
+      cephadm_bootstrap:
+        mon_ip: "{{ mon_ip }}"
+        dashboard: true
+        monitoring: false
```
by default monitoring is set to `true`. This test case is to validate bootstraping cluster with dashboard. We have separate testcase for monitoring.

vamahaja commented on playbooks/bootstrap-cluster.yml:
```diff
@@ -0,0 +1,15 @@
+  any_errors_fatal: true
+
+  tasks:
+    - name: Bootstrap with dashboard set to `true` without credenatials
```
Yes, dashboard credentials will come by deafult but in `cephadm_bootstrap` ansible module you can provide these credentails as well. We have separate test case for this scenario hence using above description for task.

vamahaja commented on suites/pacific/cephadm/test_cephadm_ansible_wrapper.yaml:
```diff
@@ -0,0 +1,54 @@
+                custom_repo: "cdn"
+                registry-url: registry.redhat.io
+                mon-ip: node1
+                skip-monitoring-stack: true
```
@sunilangadi2 thanks for suggestion. I was not sure whether deployment work without this parameter hence kept this parameter. Will update it now.

sunilangadi2 replied:
If you are going to add `Monitoring stack` separately in some other test then it's ok to keep it.

Ameenasuhani commented on tests/cephadm/test_cephadm_ansible_wrapper.py:
```diff
@@ -0,0 +1,72 @@
+CEPHADM_ANSIBLE_PATH = "/usr/share/cephadm-ansible"
+
+
+def _get_node_ip(ceph_cluster, node_name):
```
As per our discussion,
1. please modify _get_node_ip(ceph_cluster, node_name) module  such that **node_name** parameter takes dictionary 
2. _get_node_ip **could** be moved to ceph.py for re-usability purpose

Ameenasuhani commented:
@psathyan With current approach [suites/pacific/cephadm/test_cephadm_ansible_wrapper.yaml](https://github.com/red-hat-storage/cephci/pull/1724/files#diff-afd8cfee7dd4f280260fb6b3720e98a84302e67315b94fe590b608fa22391ede) we are calling "test_cephadm.py" just for enabling repos  can we make the same changes in this file [tests/cephadm/test_cephadm_ansible_wrapper.py](https://github.com/red-hat-storage/cephci/pull/1724/files#diff-95037faa5084c60157b90235947744db471ff8f476472b0a4f04d239d0216e36) and avoid tweaking bootstrapy and test_cephadm?

psathyan commented on ceph/ceph_admin/bootstrap.py:
```diff
@@ -185,6 +185,7 @@ def bootstrap(self: CephAdmProtocol, config: Dict):
                     mon-ip: <node-name>
         """
         self.cluster.setup_ssh_keys()
+        skip_cephadm_bootstrap = config.get("skip_cephadm_bootstrap", False)
```
IMO, we could decouple a couple of methods or create a new module and then transition into it. Adding this flag wouldn't be the right step working towards the goal.

psathyan commented on suites/pacific/cephadm/test_cephadm_ansible_wrapper.yaml:
```diff
@@ -0,0 +1,54 @@
+  - test:
+      name: Cluster deployment
+      desc: Cluster deployment with latest released version of 5.x
+      module: test_cephadm.py
```
IMO, we should create a separate test module that caters to this purpose. We would be expanding on this functionality.

vamahaja commented:
Test Run - https://ceph-downstream-jenkins-csb-storage.apps.ocp-c1.prod.psi.redhat.com/view/RHCS%20QE/job/rhceph-test-executor/1392/console 

Results - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-45JWJ2/ 

vamahaja commented:
Verified. Test Results - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-L0APM3/

vamahaja commented on tests/cephadm/test_cephadm_ansible_wrapper.py:
```diff
@@ -0,0 +1,104 @@
+    nodes = ceph_cluster.get_nodes()
+    installer = ceph_cluster.get_ceph_object("installer")
+
+    playbook = config.get("playbook")
```
We added check for this variable in next step. This will enable us to raise custom exception instead of Python in built exceptions.

vamahaja commented on utility/install_prereq.py:
```diff
@@ -0,0 +1,225 @@
+from utility.log import Log
+from utility.utils import fetch_build_artifacts
+
+RHEL8_EPEL_RELESE_RPM = (
+    "https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm"
+)
```
Will take this in improving install module activity.

vamahaja commented on cli/utilities/packages.py:
```diff
@@ -0,0 +1,49 @@
+    return stdout
+
+
+def yum_install(node, pkg):
```
moved this code to `packages.py` which has interfaaces for `yum/dnf`, `subscription-manager` and `rpm`.

vamahaja commented on cli/utilities/repos.py:
```diff
@@ -0,0 +1,50 @@
+def add(node, repo_url):
```
Added this support in `package.py` module.

vamahaja commented on cli/utilities/repos.py:
```diff
@@ -0,0 +1,50 @@
+
+    repos = list(map(lambda repo: f"--enable={repo}", repos))
+    cmd = f"subscription-manager repos {' '.join(repos)}"
+    stdout, stderr = node.exec_command(sudo=True, check_ec=False, cmd=cmd)
```
Made necessary changes as suggested in other modules

vamahaja commented:
Intergrated changes as discused.
Test Run - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1FYUAL/

sunilkumarn417 commented on tests/cephadm/test_cephadm_ansible_wrapper.py:
```diff
@@ -0,0 +1,110 @@
+            name: Bootstrap cluster using cephadm-ansible playbook
+            desc: Execute playbooks/bootstrap-cluster.yaml
+            config:
+                playbook: bootstrap-cluster.yaml
```
`    args = kwargs.get("config", {}).get("args")`

please add example of the args.

vamahaja replied:
Updated Testsuite yaml file config with appropriate example

sunilkumarn417 commented on tests/cephadm/test_cephadm_ansible_wrapper.py:
```diff
@@ -0,0 +1,110 @@
+            extra_vars = {}
+        extra_vars["mon_ip"] = get_node_ip(nodes, mon_node)
+
+    if Rpm(installer).query("cephadm"):
```
I think we need a arg check to differentiate bootstrap / orch or other playbooks.

for example.,
   Bootstrap --> pre-flight --> bootstrap the first mon-mgr node.

vamahaja replied:
Added new parameter `ansible_wrapper` in testsuite config

vamahaja commented on tests/cephadm/test_cephadm_ansible_wrapper.py:
```diff
@@ -0,0 +1,110 @@
+            extra_vars = {}
+        extra_vars["mon_ip"] = get_node_ip(nodes, mon_node)
+
+    if Rpm(installer).query("cephadm"):
```
Added steps to check for `cephadm-ansible` and then do required minimum steps to bootstrap node.

vamahaja commented:
Addressed comments related to cephadm and cephadm-ansible installation. Updated test configs accordingly.
Test Result - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-Z4OIXT/

sunilkumarn417 reviewed (commented):
LGTM.

There are few minor things which can consider as we progress with cephadm-ansible playbook(s),
- `setup_ssh_keys` may be configured again, if wrapper test module called multiple times.
- Currently `ansible_wrapper` doesn't consider upgrade scenarios, Since this query `if Rpm(installer).query("cephadm-ansible"):` returns `True` if cephadm-ansible already exists. it will skip, 
   - Enabling new repos
   - New cephadm-ansible installation
   - Running pre-flight.
-  May need changes w.r.t `ceph orch upgrade` playbooks, Since everything is `ceph orch` options.

vamahaja commented:
> 
Please check below comments - 

- `setup_ssh_keys` may be configured again, if wrapper test module called multiple times.
        `setup_ssh_keys` is going to be executed only one time when we are using `cephadm_bootstrap` module. Appropriate check has added for same. Rest of the test cases in this test suite would be scenarios based on same bootstraped cluster.
- Currently ansible_wrapper doesn't consider upgrade scenarios, ......
        cephadm-ansible wrapper doesn't support updagrade scenario for now. It only include `ceph_orch_host`, `ceph_orch_daemon` and `ceph_orch_apply`. If in future it add support for upgrade, will add separate test module for same.

psathyan commented on cli/utilities/packages.py:
```diff
@@ -0,0 +1,181 @@
+        if force:
+            cmd += " --force"
+
+        return self.execute(sudo=True, long_running=True, cmd=cmd)
```
IMO, we shouldn't set `long_running=True`... We need to ensure it passes and raise on failure. This being a utility and not something we are going to test... we can add intelligence here.

The implemented approach is useful when we are testing the functionality of it.

vamahaja replied:
Thanks @psathyan, updated all related methods

psathyan commented on cli/utilities/packages.py:
```diff
@@ -0,0 +1,181 @@
+            pkg (str): package need to be installed
+        """
+        cmd = f"{self.manager} install -y {pkg}"
+        return self.execute(sudo=True, long_running=True, cmd=cmd)
```
Can we handle the outcome and raise based on the intent?

vamahaja replied:
Updated all related functions

psathyan commented on cli/utilities/packages.py:
```diff
@@ -0,0 +1,181 @@
+            repo (str): repo to be enabled
+        """
+        cmd = f"yum-config-manager --add-repo {repo}"
+        if nogpgcheck:
```
I don't think we could add a repo and set `--nogpgcheck` at the same time. Please confirm.

This being a utility, I strongly believe we should handle the outcome.

psathyan commented on cli/utilities/packages.py:
```diff
@@ -0,0 +1,181 @@
+        """
+        cmd = f"{self.base_cmd} --query {pkg}"
+
+        return self.execute(sudo=True, long_running=True, cmd=cmd)
```
My two cents on this module... you have rightly classified it as a utility hence, handling the unexpected behavior needs to occur within this module. On event of such behavior an exception must be raised.

vamahaja replied:
done

psathyan commented on cli/utilities/utils.py:
```diff
@@ -28,11 +28,20 @@ def get_running_containers(node, expr=None, **kw):
+    Args:
+        node (ceph): Ceph node object
+    """
+    cmd = r"cat /etc/os-release | tr -dc '0-9.'| cut -d \. -f1"
```
I always found this useful `eval $(cat /etc/os-release) ; echo $VERSION_ID`

vamahaja replied:
@psathyan `VERSION_ID` returns `major.minor` version. If we need to use above command then it should be `eval $(cat /etc/os-release) ; echo $VERSION_ID | cut -d \. -f1`

psathyan reviewed (commented):
Looks good to me

manasagowri reviewed (approved):
LGTM.
